File: packages/core/src/retry.ts
// packages/core/src/retry.ts
// Retry with exponential backoff + Circuit breaker

import { createLogger } from './logger.js'

const log = createLogger('core:retry')

export interface RetryOptions {
  maxRetries?: number
  baseDelayMs?: number
  maxDelayMs?: number
  retryOn?: (error: unknown) => boolean
}

const DEFAULT_RETRY: Required<RetryOptions> = {
  maxRetries: 3,
  baseDelayMs: 1000,
  maxDelayMs: 30_000,
  retryOn: (error: unknown) => {
    if (error instanceof Error) {
      const msg = error.message.toLowerCase()
      // Retry on rate limit, timeout, server errors
      return msg.includes('rate') || msg.includes('429') ||
             msg.includes('timeout') || msg.includes('500') ||
             msg.includes('502') || msg.includes('503') ||
             msg.includes('overloaded')
    }
    return false
  },
}

export async function withRetry<T>(
  fn: () => Promise<T>,
  opts: RetryOptions = {},
): Promise<T> {
  const { maxRetries, baseDelayMs, maxDelayMs, retryOn } = { ...DEFAULT_RETRY, ...opts }

  let lastError: unknown
  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      return await fn()
    } catch (error) {
      lastError = error
      if (attempt >= maxRetries || !retryOn(error)) {
        throw error
      }
      const delay = Math.min(baseDelayMs * Math.pow(2, attempt) + Math.random() * 500, maxDelayMs)
      log.warn(`Retry ${attempt + 1}/${maxRetries} after ${Math.round(delay)}ms: ${error instanceof Error ? error.message : error}`)
      await new Promise(r => setTimeout(r, delay))
    }
  }
  throw lastError
}

// --- Circuit Breaker ---

export type CircuitState = 'closed' | 'open' | 'half-open'

export interface CircuitBreakerOptions {
  failureThreshold?: number
  resetTimeoutMs?: number
}

export class CircuitBreaker {
  private state: CircuitState = 'closed'
  private failureCount = 0
  private lastFailureTime = 0
  private failureThreshold: number
  private resetTimeoutMs: number

  constructor(private name: string, opts: CircuitBreakerOptions = {}) {
    this.failureThreshold = opts.failureThreshold ?? 5
    this.resetTimeoutMs = opts.resetTimeoutMs ?? 60_000
  }

  async execute<T>(fn: () => Promise<T>): Promise<T> {
    if (this.state === 'open') {
      if (Date.now() - this.lastFailureTime > this.resetTimeoutMs) {
        this.state = 'half-open'
        log.info(`Circuit ${this.name}: half-open, trying one request`)
      } else {
        throw new Error(`Circuit breaker ${this.name} is OPEN. Try again later.`)
      }
    }

    try {
      const result = await fn()
      if (this.state === 'half-open') {
        this.state = 'closed'
        this.failureCount = 0
        log.info(`Circuit ${this.name}: closed (recovered)`)
      }
      return result
    } catch (error) {
      this.failureCount++
      this.lastFailureTime = Date.now()
      if (this.failureCount >= this.failureThreshold) {
        this.state = 'open'
        log.error(`Circuit ${this.name}: OPEN after ${this.failureCount} failures`)
      }
      throw error
    }
  }

  getState(): CircuitState { return this.state }
  reset(): void { this.state = 'closed'; this.failureCount = 0 }
}

File: crates/quasar-core/src/retry.rs
use std::future::Future;
use std::sync::Arc;
use std::time::{Duration, Instant};
use parking_lot::RwLock;
use tokio::time::sleep;
use crate::{QuasarError, QuasarResult};

#[derive(Debug, Clone)]
pub struct RetryOptions {
    pub max_retries: usize,
    pub base_delay_ms: u64,
    pub max_delay_ms: u64,
}

impl Default for RetryOptions {
    fn default() -> Self {
        Self {
            max_retries: 3,
            base_delay_ms: 1000,
            max_delay_ms: 30_000,
        }
    }
}

pub async fn with_retry<F, Fut, T>(mut f: F, options: RetryOptions) -> QuasarResult<T>
where
    F: FnMut() -> Fut,
    Fut: Future<Output = QuasarResult<T>>,
{
    let mut attempt = 0;
    loop {
        match f().await {
            Ok(result) => return Ok(result),
            Err(e) => {
                attempt += 1;
                if attempt > options.max_retries {
                    return Err(e);
                }
                let delay = std::cmp::min(
                    options.base_delay_ms * 2_u64.pow(attempt as u32 - 1),
                    options.max_delay_ms,
                );
                tracing::warn!("Retry attempt {}/{} after {}ms: {:?}", attempt, options.max_retries, delay, e);
                sleep(Duration::from_millis(delay)).await;
            }
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CircuitState {
    Closed,
    Open,
    HalfOpen,
}

pub struct CircuitBreaker {
    name: String,
    state: Arc<RwLock<CircuitState>>,
    failure_count: Arc<RwLock<usize>>,
    last_failure_time: Arc<RwLock<Option<Instant>>>,
    failure_threshold: usize,
    reset_timeout: Duration,
}

impl Clone for CircuitBreaker {
    fn clone(&self) -> Self {
        Self {
            name: self.name.clone(),
            state: Arc::clone(&self.state),
            failure_count: Arc::clone(&self.failure_count),
            last_failure_time: Arc::clone(&self.last_failure_time),
            failure_threshold: self.failure_threshold,
            reset_timeout: self.reset_timeout,
        }
    }
}

impl CircuitBreaker {
    pub fn new(name: impl Into<String>, failure_threshold: usize, reset_timeout: Duration) -> Self {
        Self {
            name: name.into(),
            state: Arc::new(RwLock::new(CircuitState::Closed)),
            failure_count: Arc::new(RwLock::new(0)),
            last_failure_time: Arc::new(RwLock::new(None)),
            failure_threshold,
            reset_timeout,
        }
    }

    pub fn state(&self) -> CircuitState {
        *self.state.read()
    }

    pub fn check_state(&self) {
        let mut state = self.state.write();
        if *state == CircuitState::Open {
            if let Some(last_failure) = *self.last_failure_time.read() {
                if last_failure.elapsed() >= self.reset_timeout {
                    tracing::info!("Circuit breaker '{}' transitioning to HalfOpen", self.name);
                    *state = CircuitState::HalfOpen;
                    *self.failure_count.write() = 0;
                }
            }
        }
    }

    pub async fn execute<F, Fut, T>(&self, f: F) -> QuasarResult<T>
    where
        F: FnOnce() -> Fut,
        Fut: Future<Output = QuasarResult<T>>,
    {
        self.check_state();

        let state = self.state();
        if state == CircuitState::Open {
            return Err(QuasarError::provider(format!(
                "Circuit breaker '{}' is OPEN",
                self.name
            )));
        }

        match f().await {
            Ok(result) => {
                if state == CircuitState::HalfOpen {
                    tracing::info!("Circuit breaker '{}' transitioning to Closed", self.name);
                    *self.state.write() = CircuitState::Closed;
                }
                *self.failure_count.write() = 0;
                Ok(result)
            }
            Err(e) => {
                let mut count = self.failure_count.write();
                *count += 1;
                *self.last_failure_time.write() = Some(Instant::now());

                if *count >= self.failure_threshold {
                    tracing::error!("Circuit breaker '{}' transitioning to OPEN after {} failures", self.name, count);
                    *self.state.write() = CircuitState::Open;
                }

                Err(e)
            }
        }
    }
}
